File: 2024/dec03/02.py
inp = open('input.txt', 'r')
all = inp.read()

def mul(n1,n2):
    return n1 * n2

totalCount = 0
do = True
for i in range(0,len(all)):
    if all[i:i+4]=='do()':
        do = True
    if all[i:i+7]=="don't()":
        do = False
    if all[i:i+3]=='mul' and do: #Estamos perante uma multiplicação e podemos resolver
        for i2 in range(3,13): #Seq à frente
            try:
                totalCount+= eval(all[i:i+i2])
            except:
                continue
            else:
                break

print('Sum of enabled multiplications (challenge 2):', totalCount)
#Right answer: 74838033

File: 2024/dec03/02.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#define BUFMAX 32768

int mul(int n1, int n2) {
    return n1*n2;
}

int main() {
    FILE *input;
    char file[BUFMAX]={0};
    char doExec[] = "do()";
    char doNotExec[] = "don't()";
    int totalCount=0;
    int n1,n2,charsRead;
    bool exec = true;
    
    input=fopen("input.txt","r");
    if (input==NULL) return 1;
    fread(file,1,BUFMAX,input);

    for (char *filePos = file; filePos < file+BUFMAX; filePos++) {
        if (memcmp(filePos,doExec,4)==0)
            exec = true;
        else if (memcmp(filePos,doNotExec,7)==0)
            exec = false;
        else if (sscanf(filePos,"mul(%d,%d)%n",&n1,&n2,&charsRead) == 2) {
            if (filePos[charsRead-1] == ')' && exec) {
                totalCount+=mul(n1,n2);
            } 
        }
    }

    printf("Sum of enabled multiplications (challenge 2): %d\n",totalCount);
    //Right answer: 74838033
    fclose(input);
    return 0;
}